### PR TITLE
[CMake] Add SSE definitions for SSE 4.1 and 4.2

### PR DESCRIPTION
**Pull request by larshg**
To enable compilation of code that is wrapped in `#if defined __SSE_4_X__`

### CODE REVIEW

**Comment by mvieth** [comment]
What about `SSE3`? Not sure if we need that, but for completeness?

**Comment by larshg** [comment]
Ah, yes. I'll add it.

**Review by mvieth — APPROVED** [review]

**Comment by SunBlack** [comment]
@larshg : Can you adjust this MR like the AVX-MR? So `string(APPEND ...)` and `function` instead of `macro`?

**Comment by larshg** [comment]
Sure I'll updated it.

File: cmake/pcl_find_sse.cmake [hunk 1]
```diff
@@ -1,210 +1,254 @@
...
+      list(APPEND SSE_FLAGS "-ffloat-store")
+    endif()
+  elseif(MSVC AND NOT CMAKE_CL_64)
+    if(SSE_LEVEL GREATER_EQUAL 2.0)
+      set( SSE_FLAGS "/arch:SSE2")
+    elseif(SSE_LEVEL GREATER_EQUAL 1.0)
+      set( SSE_FLAGS "/arch:SSE")
     endif()
+  elseif(MSVC)
+    if(SSE_LEVEL GREATER_EQUAL 4.2)
```
Comment (kunaltyagi) [inline cmake/pcl_find_sse.cmake:230]: So this is for Clang compiler on Windows?

Comment (larshg) [reply cmake/pcl_find_sse.cmake:230]: No, its only for MSVC 64 bit. It doesn't look at the OS, only at which compiler. Clang is part of line 205.

Comment (kunaltyagi) [reply cmake/pcl_find_sse.cmake:230]: So MSVC 64 and 32 bit have different flags? That's pretty awesome.

File: cmake/pcl_find_sse.cmake [hunk 2]
```diff
@@ -1,210 +1,254 @@
...
+    HAVE_MM_MALLOC)
+
+  check_cxx_source_runs("
+    #include <stdlib.h>
+    int main()
+    {
+      void* mem;
+      return posix_memalign (&mem, 16, 100);
+    }"
+    HAVE_POSIX_MEMALIGN)
```
Comment (kunaltyagi) [inline cmake/pcl_find_sse.cmake:53]: We should remove this soon. The only place this is used is in linemod (recognition)

**Review by kunaltyagi — APPROVED** [review]
Looks good(-ish) to me (not an expert in either CMake or sse)

**Comment by larshg** [comment]
@SunBlack Would you have a look 😄  ?

File: cmake/pcl_find_sse.cmake [hunk 3]
```diff
@@ -1,210 +1,254 @@
...
+      set(SSE_FLAGS "-msse4.1 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.1)
+      set(SSE_FLAGS "-mssse3 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.0)
+      set(SSE_FLAGS "-msse3 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 2.0)
+      set(SSE_FLAGS "-msse2 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 1.0)
+      set(SSE_FLAGS "-msse -mfpmath=sse")
+    else()
```
Comment (SunBlack) [inline cmake/pcl_find_sse.cmake:218]: Just a matter of taste.

Maybe you want to deduplicate the `-mfpmath=sse`:

```suggestion
  if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
    if(SSE_LEVEL GREATER_EQUAL 1.0)
      if(SSE_LEVEL GREATER_EQUAL 4.2)
        set(SSE_FLAGS "-msse4.2")
      elseif(SSE_LEVEL GREATER_EQUAL 4.1)
        set(SSE_FLAGS "-msse4.1")
      elseif(SSE_LEVEL GREATER_EQUAL 3.1)
        set(SSE_FLAGS "-msse3")
      elseif(SSE_LEVEL GREATER_EQUAL 3.0)
        set(SSE_FLAGS "-msse3")
      elseif(SSE_LEVEL GREATER_EQUAL 2.0)
        set(SSE_FLAGS "-msse2")
      else()
        set(SSE_FLAGS "-msse")
      endif()
      string(APPEND SSE_FLAGS " -mfpmath=sse")
    else()
```

But current variant is also okay for me.

Another possibility would be to use `set(SSE_FLAGS "-msse${SSE_LEVEL} -mfpmath=sse")` for SSE2 or greater (in this case you should change `set(SSE_LEVEL 2.0)` to `set(SSE_LEVEL 2)`). This is short, but less readable (so imho current version is maybe better).

File: cmake/pcl_find_sse.cmake [hunk 4]
```diff
@@ -1,210 +1,254 @@
...
+    endif()
+  endif()
+  
+  if(CMAKE_COMPILER_IS_GNUCC OR CMAKE_COMPILER_IS_GNUCXX OR CMAKE_COMPILER_IS_CLANG)
+    if(SSE_LEVEL GREATER_EQUAL 4.2)
+      set(SSE_FLAGS "-msse4.2 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 4.1)
+      set(SSE_FLAGS "-msse4.1 -mfpmath=sse")
+    elseif(SSE_LEVEL GREATER_EQUAL 3.1)
+      set(SSE_FLAGS "-mssse3 -mfpmath=sse")
```
Comment (SunBlack) [inline cmake/pcl_find_sse.cmake:211]: I believe ;-)

```suggestion
      set(SSE_FLAGS "-msse3 -mfpmath=sse")
```

File: cmake/pcl_find_sse.cmake [hunk 5]
```diff
@@ -1,210 +1,254 @@
...
+    if(SSE_LEVEL GREATER_EQUAL 3.1)
+      string(APPEND SSE_DEFINITIONS " -D__SSSE3__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 3.0)
+      string(APPEND SSE_DEFINITIONS " -D__SSE3__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 2.0)
+      string(APPEND SSE_DEFINITIONS " -D__SSE2__")
+    endif()
+    if(SSE_LEVEL GREATER_EQUAL 1.0)
```
Comment (SunBlack) [inline cmake/pcl_find_sse.cmake:245]: For Gnu you are using `elseif()`. Isn't the definition of the latest version enough?

File: cmake/pcl_find_sse.cmake [hunk 6]
```diff
@@ -1,210 +1,254 @@
...
-        if(HAVE_SSE2_EXTENSIONS)
-            string(APPEND SSE_DEFINITIONS " -D__SSE2__")
-        endif()
-        if(HAVE_SSE_EXTENSIONS)
-            string(APPEND SSE_DEFINITIONS " -D__SSE__")
-        endif()
-    endif()
-    string(REPLACE ";" " " SSE_FLAGS_STR "${SSE_FLAGS}")
-endmacro()
+  unset(CMAKE_REQUIRED_FLAGS)
```
Comment (SunBlack) [inline cmake/pcl_find_sse.cmake:253]: It would be saver to store the value at start and reset it instead of just to unset it

**Review by SunBlack — CHANGES_REQUESTED** [review]
Another review iteration done ;-)

File: cmake/pcl_find_sse.cmake (hunk 4 above)
Comment (larshg) [reply cmake/pcl_find_sse.cmake:211]: Its with 3 sss's, see https://en.wikipedia.org/wiki/SSSE3

SSE3 is just below.

File: cmake/pcl_find_sse.cmake (hunk 6 above)
Comment (larshg) [reply cmake/pcl_find_sse.cmake:256]: I think is mainly used for temporary flags used with `check_cxx_source_runs`. Current findSSE just set it to empty like `set(CMAKE_REQUIRED_FLAGS)`

File: cmake/pcl_find_sse.cmake (hunk 5 above)
Comment (larshg) [reply cmake/pcl_find_sse.cmake:248]: For Gnu its the compiler flag.
And I by the test I have done, the more secure way is to add all the definitions that are available. Some definitions doesn't seem to be set, if only the compiler flags are used. And therefore some handcraftet guarded code, won't be compiled.

Comment (larshg) [reply cmake/pcl_find_sse.cmake:248]: I'm actually not fully sure, if this is also the case for linux. I have mainly tested on Windows.

File: cmake/pcl_find_sse.cmake (hunk 4 above)
Comment (SunBlack) [reply cmake/pcl_find_sse.cmake]: Ahh good to know.

File: cmake/pcl_find_sse.cmake (hunk 5 above)
Comment (SunBlack) [reply cmake/pcl_find_sse.cmake:248]: You could try  it on godbolt ;-)

File: cmake/pcl_find_sse.cmake (hunk 6 above)
Comment (SunBlack) [reply cmake/pcl_find_sse.cmake:256]: Okay checked the CMake sources and the variable seems never to be set from CMake itself. So it is okay.

**Comment by kunaltyagi** [comment]
🚀 ?

File: cmake/pcl_find_sse.cmake [hunk 7]
```diff
@@ -1,210 +1,257 @@
...
+      else()
+        set(SSE_FLAGS "-msse")
+      endif()
+      string(APPEND SSE_FLAGS " -mfpmath=sse")
+    else()
+      # Setting -ffloat-store to alleviate 32bit vs 64bit discrepancies on non-SSE
+      # platforms.
+      list(APPEND SSE_FLAGS "-ffloat-store")
+    endif()
+  elseif(MSVC AND NOT CMAKE_CL_64)
```
Comment (SunBlack) [inline cmake/pcl_find_sse.cmake:226]: Just saw in latest [documentation](https://cmake.org/cmake/help/latest/variable/CMAKE_CL_64.html), that `CMAKE_CL_64` is deprecated and the alternate is already supported by CMake 3.5.

Comment (larshg) [reply cmake/pcl_find_sse.cmake:226]: Ahh, I'll update it later.

**Pull request merged**